File: FHD/frontend/src/utils/employeeSsotContacts.ts
import { YUANGON_PKG_ROLE_LABELS } from '@/domain/yuangonDutyRoster'

export type EmployeeSsotContactRecord = {
  employee_id: string
  display_name: string
  surface_name?: string
  department?: string
  source?: string
  installed?: boolean
  runnable?: boolean
  online?: boolean
  pinned?: boolean
  avatar_key?: string
  contact_route?: string
  mobile_contact_route?: string
  capabilities?: string[]
  last_task_status?: string
  description?: string
}

export type EmployeeSsotContact = {
  id: string
  display_name: string
  username: string
  subtitle: string
  description: string
  area: string
  status: string
  api_base_path: string
  phone_channel: string
  is_duty_employee_entry: true
  source?: string
  installed?: boolean
  runnable?: boolean
  pinned?: boolean
}

export type EmployeeSsotPayload = {
  contacts?: EmployeeSsotContactRecord[]
  employee_labels?: Record<string, string>
  employee_descriptions?: Record<string, string>
  admin?: {
    departments?: Array<{
      id?: string
      key?: string
      label?: string
      employees?: Array<{ id?: string; on_duty?: boolean | null }>
    }>
    on_duty_employee_ids?: string[] | null
  }
  enterprise?: {
    layers?: Array<{ id?: string; label?: string }>
    employees?: Record<
      string,
      {
        id?: string
        label?: string
        listing?: string
        enterprise_layer?: string
        description?: string
      }
    >
  }
}

/** 客户可读员工名：后端真实名优先；后端回显英文 ID 时用 SSOT label / 编制表中文名兜底。 */
function nameForEmployee(
  id: string,
  row: EmployeeSsotContactRecord | null,
  payload: EmployeeSsotPayload | null | undefined,
): string {
  const backendName = String(row?.display_name || '').trim()
  if (backendName && backendName !== id) return backendName
  return String(payload?.employee_labels?.[id] || YUANGON_PKG_ROLE_LABELS[id] || backendName || id).trim() || id
}

function descriptionForEmployee(id: string, payload: EmployeeSsotPayload | null | undefined, fallback = ''): string {
  const descriptions = payload?.employee_descriptions || {}
  return String(descriptions[id] || fallback).trim()
}

/** 部门/层级英文 ID → 客户可读的中文部门名（管理端六线 label 或企业端四层 label）。 */
function areaLabelFor(
  id: string,
  row: EmployeeSsotContactRecord,
  payload: EmployeeSsotPayload | null | undefined,
): string {
  const raw = String(row.department || '').trim()
  const admin = payload?.admin?.departments || []
  for (const dept of admin) {
    if (raw && (dept.id === raw || dept.key === raw)) {
      return String(dept.label || dept.id || '').trim() || '编制'
    }
  }
  const layers = payload?.enterprise?.layers || []
  const layerLabel = (layerId: string): string => {
    for (const layer of layers) {
      if (layer.id === layerId) return String(layer.label || layer.id || '').trim() || '编制'
    }
    return ''
  }
  if (raw) {
    const byDept = layerLabel(raw)
    if (byDept) return byDept
  }
  const entLayer = String(payload?.enterprise?.employees?.[id]?.enterprise_layer || '').trim()
  if (entLayer) return layerLabel(entLayer) || '编制'
  return '编制'
}

function mapContactRecord(row: EmployeeSsotContactRecord, payload: EmployeeSsotPayload | null | undefined): EmployeeSsotContact | null {
  const id = String(row.employee_id || '').trim()
  if (!id) return null
  const displayName = nameForEmployee(id, row, payload)
  const areaLabel = areaLabelFor(id, row, payload)
  const installed = Boolean(row.installed)
  const runnable = Boolean(row.runnable ?? installed)
  const source = String(row.source || (installed ? 'installed' : 'planned')).trim()
  const contactRoute = String(row.contact_route || row.mobile_contact_route || `/api/admin/employees/${id}`).trim()
  return {
    id,
    display_name: displayName,
    username: id,
    subtitle: runnable ? `${areaLabel} · 可执行` : `${areaLabel} · 未安装`,
    description:
      String(row.description || '').trim() ||
      descriptionForEmployee(id, payload, runnable ? '已安装，可联系' : '编制内但未安装员工包'),
    area: areaLabel,
    status: runnable ? 'on_duty' : 'planned',
    api_base_path: contactRoute.replace(/\/chat\/?$/, '').replace(/\/messages\/?$/, '') || `/api/admin/employees/${id}`,
    phone_channel: source.startsWith('codex') || source === 'builtin' ? 'super' : 'admin-duty',
    is_duty_employee_entry: true,
    source,
    installed,
    runnable,
    pinned: Boolean(row.pinned),
  }
}

/** 将 platform-shell / mobile employee-ssot 派生包转为 IM 侧栏联系人。 */
export function dutyEmployeesFromEmployeeSsot(payload: EmployeeSsotPayload | null | undefined): EmployeeSsotContact[] {
  const contacts = payload?.contacts
  if (Array.isArray(contacts) && contacts.length) {
    const out: EmployeeSsotContact[] = []
    const seen = new Set<string>()
    for (const row of contacts) {
      if (!row || typeof row !== 'object') continue
      const source = String(row.source || '').trim()
      if (source === 'builtin' || source === 'codex') continue
      const mapped = mapContactRecord(row, payload)
      if (!mapped || seen.has(mapped.id)) continue
      seen.add(mapped.id)
      out.push(mapped)
    }
    return out.sort((a, b) => {
      if (Boolean(a.runnable) !== Boolean(b.runnable)) {
        return a.runnable ? -1 : 1
      }
      return a.display_name.localeCompare(b.display_name, 'zh-CN')
    })
  }

  if (!payload?.admin?.departments?.length) return []
  const out: EmployeeSsotContact[] = []
  const onDutySet = new Set(payload.admin.on_duty_employee_ids || [])
  for (const dept of payload.admin.departments) {
    const area = String(dept.label || dept.id || '').trim() || '编制'
    for (const emp of dept.employees || []) {
      const id = String(emp.id || '').trim()
      if (!id) continue
      const onDuty = emp.on_duty ?? onDutySet.has(id)
      out.push({
        id,
        display_name: nameForEmployee(id, null, payload),
        username: id,
        subtitle: onDuty ? `${area} · 可执行` : `${area} · 未安装`,
        description: descriptionForEmployee(id, payload, onDuty ? '已安装，可联系' : '编制内但未安装员工包'),
        area,
        status: onDuty ? 'on_duty' : 'planned',
        api_base_path: `/api/admin/employees/${id}`,
        phone_channel: 'admin-duty',
        is_duty_employee_entry: true,
        source: onDuty ? 'installed' : 'planned',
        installed: onDuty,
        runnable: onDuty,
      })
    }
  }
  return out
}

File: FHD/app/mod_sdk/employee_ssot_contacts.py
"""员工联系人 SSOT 派生（三端统一 contacts 契约）。

从 :mod:`app.mod_sdk.employee_ssot` 拆出，避免 ``employee_ssot.py`` 超过架构 fitness 上限。
"""

from __future__ import annotations

import json
from functools import lru_cache
from typing import Any

from app.mod_sdk.host_profile import resolve_fhd_config_dir
from app.utils.operational_errors import RECOVERABLE_ERRORS

SUPER_EMPLOYEE_CONTACT_ORDER: tuple[str, ...] = (
    "codex-super-employee",
    "cursor-super-employee",
    "claude-super-employee",
    "trae-super-employee",
)


@lru_cache(maxsize=1)
def load_employee_manifest_meta() -> dict[str, dict[str, str]]:
    """从 ``mods/_employees/*/manifest.json`` 派生 ``{emp_id: {name, description}}``。"""
    out: dict[str, dict[str, str]] = {}
    cfg_dir = resolve_fhd_config_dir()
    if cfg_dir is None:
        return out
    employees_dir = cfg_dir.parent / "mods" / "_employees"
    if not employees_dir.is_dir():
        return out
    for manifest_path in employees_dir.glob("*/manifest.json"):
        try:
            data = json.loads(manifest_path.read_text(encoding="utf-8"))
            if not isinstance(data, dict):
                continue
            emp_id = str(data.get("id") or "").strip()
            if not emp_id:
                continue
            raw_employee_meta = data.get("employee")
            employee_meta: dict[str, Any] = (
                dict(raw_employee_meta) if isinstance(raw_employee_meta, dict) else {}
            )
            name = str(data.get("name") or employee_meta.get("label") or "").strip()
            desc = str(data.get("description") or "").strip()
            row: dict[str, str] = {}
            if name:
                row["name"] = name
            if desc:
                row["description"] = desc
            if row:
                out[emp_id] = row
        except RECOVERABLE_ERRORS:
            continue
    return out


def employee_display_name(emp_id: str) -> str:
    meta = load_employee_manifest_meta().get(str(emp_id or "").strip()) or {}
    return str(meta.get("name") or emp_id).strip() or emp_id


def employee_description(emp_id: str) -> str:
    meta = load_employee_manifest_meta().get(str(emp_id or "").strip()) or {}
    return str(meta.get("description") or "").strip()


def _employee_contact_record(
    employee_id: str,
    *,
    display_name: str,
    department: str,
    source: str,
    installed: bool,
    pinned: bool = False,
    surface_name: str | None = None,
    description: str = "",
    capabilities: list[str] | None = None,
    contact_route: str | None = None,
    mobile_contact_route: str | None = None,
    last_task_status: str = "idle",
) -> dict[str, Any]:
    eid = str(employee_id or "").strip()
    runnable = bool(installed and source in {"installed", "builtin", "codex"})
    avatar_key = eid.split("-")[0][:24] if eid else "employee"
    return {
        "employee_id": eid,
        "display_name": display_name,
        "surface_name": surface_name or display_name,
        "department": department,
        "source": source,
        "installed": installed,
        "runnable": runnable,
        "online": runnable,
        "pinned": pinned,
        "avatar_key": avatar_key,
        "contact_route": contact_route or f"/api/admin/employees/chat/{eid}",
        "mobile_contact_route": mobile_contact_route or f"/api/mobile/v1/employees/{eid}/messages",
        "capabilities": list(capabilities or []),
        "last_task_status": last_task_status,
        "description": description,
    }


def _super_employee_contacts() -> list[dict[str, Any]]:
    from app.mod_sdk import assistant_ssot

    registry = assistant_ssot.super_employees()
    out: list[dict[str, Any]] = []
    for emp_id in SUPER_EMPLOYEE_CONTACT_ORDER:
        meta = registry.get(emp_id)
        if not isinstance(meta, dict):
            continue
        if assistant_ssot.is_factory_employee(emp_id):
            continue
        display = str(meta.get("display_name") or emp_id).strip()
        summary = str(
            meta.get("summary") or f"{meta.get('display_tool') or ''} 超级员工 · 多设备派工".strip()
        ).strip()
        source = "codex" if emp_id == "codex-super-employee" else "builtin"
        out.append(
            _employee_contact_record(
                emp_id,
                display_name=display,
                department="super",
                source=source,
                installed=True,
                pinned=emp_id == "codex-super-employee",
                description=summary,
                contact_route=f"/api/admin/{emp_id}/messages",
                mobile_contact_route=f"/api/mobile/v1/admin/{emp_id}/messages",
            )
        )
    return out


def derive_employee_contacts(
    installed_ids: set[str] | frozenset[str] | None = None,
    *,
    include_super: bool = True,
    include_enterprise_listed: bool = True,
) -> list[dict[str, Any]]:
    """三端统一的员工联系人列表（Web IM / 手机消息 / AI 员工页共用）。"""
    from app.mod_sdk.employee_ssot import (
        LISTING_LISTED,
        derive_admin_duty_roster,
        load_enterprise_employees,
    )

    installed = frozenset(str(x).strip() for x in (installed_ids or ()) if str(x).strip())
    manifest_meta = load_employee_manifest_meta()
    admin = derive_admin_duty_roster(installed_ids=installed)
    on_duty = set(admin.get("on_duty_employee_ids") or [])
    seen: set[str] = set()
    contacts: list[dict[str, Any]] = []

    if include_super:
        for row in _super_employee_contacts():
            eid = str(row.get("employee_id") or "").strip()
            if not eid or eid in seen:
                continue
            seen.add(eid)
            contacts.append(row)

    for dept in admin.get("departments") or []:
        if not isinstance(dept, dict):
            continue
        department = str(dept.get("id") or dept.get("key") or dept.get("label") or "admin").strip()
        for emp in dept.get("employees") or []:
            if not isinstance(emp, dict):
                continue
            eid = str(emp.get("id") or "").strip()
            if not eid or eid in seen:
                continue
            seen.add(eid)
            meta = manifest_meta.get(eid) or {}
            display = str(meta.get("name") or eid).strip() or eid
            desc = str(meta.get("description") or "").strip()
            is_installed = eid in on_duty
            contacts.append(
                _employee_contact_record(
                    eid,
                    display_name=display,
                    department=department,
                    source="installed" if is_installed else "planned",
                    installed=is_installed,
                    description=desc
                    or ("已安装，可联系" if is_installed else "编制内但未安装 employee_pack"),
                )
            )

    if include_enterprise_listed:
        for meta in load_enterprise_employees().values():
            if meta.get("listing") != LISTING_LISTED:
                continue
            eid = str(meta.get("id") or "").strip()
            if not eid or eid in seen:
                continue
            seen.add(eid)
            display = str(
                meta.get("label") or manifest_meta.get(eid, {}).get("name") or eid
            ).strip()
            desc = str(manifest_meta.get(eid, {}).get("description") or "").strip()
            is_installed = eid in installed
            contacts.append(
                _employee_contact_record(
                    eid,
                    display_name=display or eid,
                    department=str(meta.get("enterprise_layer") or "management"),
                    source="installed" if is_installed else "planned",
                    installed=is_installed,
                    description=desc,
                )
            )

    return contacts


def employee_label_maps() -> tuple[dict[str, str], dict[str, str]]:
    manifest_meta = load_employee_manifest_meta()
    employee_labels = {
        eid: str(meta.get("name") or eid) for eid, meta in manifest_meta.items() if meta.get("name")
    }
    employee_descriptions = {
        eid: str(meta.get("description") or "")
        for eid, meta in manifest_meta.items()
        if meta.get("description")
    }
    return employee_labels, employee_descriptions


__all__ = [
    "SUPER_EMPLOYEE_CONTACT_ORDER",
    "derive_employee_contacts",
    "employee_description",
    "employee_display_name",
    "employee_label_maps",
    "load_employee_manifest_meta",
]
